File: smartCalculator.py
import json
import os

# file path where the json file will be saved
json_path = r"C:\Users\a\source\repos\homework\homework\bin\Debug\section_C_task2.json"


# checking if there is json file already
def checking_history_exists():
    if not os.path.exists(json_path):
        # if there is not, creating an empty file
        with open(json_path, 'w') as file:
            json.dump([], file)


# saving history to the json file
def saving_history(expression, result):
    checking_history_exists()
    with open(json_path, 'r') as file:
        history = json.load(file)
    
    history.append({'expression': expression, 'result': result})
    
    with open(json_path, 'w') as file:
        json.dump(history, file, indent=4)


# first option for basic calculation
def first_option():
    user_input = input("Enter your calculation: ")
    
    try:
        tokens_of_input = user_input.split(' ')

        if len(tokens_of_input) != 3:
            raise ValueError("You did not enter the valid format.")

        first_number = float(tokens_of_input[0])
        operator = tokens_of_input[1]
        second_number = float(tokens_of_input[2])
        result = 0

        # Checking user's wanted operation and updating history
        if operator == "+":
            result = first_number + second_number
            print(f"Result: {result}")
            saving_history(user_input, result)
        elif operator == "-":
            result = first_number - second_number
            print(f"Result: {result}")
            saving_history(user_input, result)
        elif operator == "*":
            result = first_number * second_number
            print(f"Result: {result}")
            saving_history(user_input, result)
        elif operator == "/":
            if second_number != 0:
                result = first_number / second_number
                print(f"Result: {result}")
                saving_history(user_input, result)
            else:
                print("Error: You cannot divide by zero.")
        else:
            print("Invalid operator. Use +, -, *, or /.")
    
    except ValueError:
        print("Invalid input format. Please use the format: operand (space) operator (space) operand.")


# second option for calculating square root
def second_option():
    user_input = input("Please enter a number to calculate its square root: ")

    try:
        user_number = float(user_input)

        if user_number < 0:
            raise ValueError("No calculations of the square root of a negative number.")

        result = user_number ** 0.5
        print(f"Square Root: √{user_number} = {result}")
        saving_history(f"√{user_number}", result)
    
    except ValueError as e:
        print(f"Error: {e}")


# option for viewing history
def third_option():
    checking_history_exists()
    with open(json_path, 'r') as file:
        history = json.load(file)

    if not history:
        print("No history available.")
    else:
        print("Calculation History:")
        for count, record in enumerate(history, start=1):
            print(f"{count}: {record['expression']} = {record['result']}")


# main function for menu and user choice
def main():
    while True:
        # starting menu
        print("Calculator Menu:")
        print("1. Perform Basic Calculation")
        print("2. Calculate Square Root")
        print("3. View History")
        print("4. Exit")
        option = input("Enter your choice: ")

        # checking if the user input is a valid number
        if option.isdigit():
            option = int(option)
            print(f"You selected option number {option}.")
        else:
            print("Invalid input. Please enter a valid number.")
            continue

        if option == 1:
            first_option()
        elif option == 2:
            second_option()
        elif option == 3:
            third_option()
        elif option == 4:
            print("Exiting the calculator.")
            break
        else:
            print("You did not enter a valid option. Try again.")


# run the calculator
main()

File: smartCalculator.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;

class Calculator
{

    // file path where the json file will be saved
    static readonly string jsonPath = @"C:\Users\a\source\repos\homework\homework\bin\Debug\section_C_task_1.json";

    // checking if there is json file already
    static void CheckingHistoryExists()
    {
        if (!File.Exists(jsonPath))
        {
            // if there is not, creating an empty file
            File.WriteAllText(jsonPath, JsonConvert.SerializeObject(new List<string>()));
        }
    }
    static void Main(string[] args)
    {
        // starting menu
        Console.WriteLine("Calculator Menu:");
        Console.WriteLine("1. Perform Basic Calculation");
        Console.WriteLine("2. Calculate Square Root");
        Console.WriteLine("3. View History");
        Console.WriteLine("4. Exit");
        Console.Write("Enter your choice: ");

        // checking user input if not valid
        if (int.TryParse(Console.ReadLine(), out int option))
        {
            Console.WriteLine($"You selected option number {option}.");
        }
        else
        {
            Console.WriteLine("Invalid input. Please enter a valid number.");
        }

        switch (option)
        {
            case 1:
                FirstOption();
                break;
            case 2:
                SecondOption();
                break;
            case 3:
                ThirdOption();
                break;
            case 4:
                Console.WriteLine("Exiting the calculator.");
                return;
            default:
                Console.WriteLine("You did not enter valid option. Try again.");
                break;

        }
    }
    static void FirstOption()
    {
        Console.Write("Enter your calculation: ");
        string userInput = Console.ReadLine();

        try
        {
            // splitting user input into 3 tokens
            string[] tokensOfInput = userInput.Split(' ');

            if (tokensOfInput.Length != 3)
            {
                throw new FormatException("You did not enter the valid format.");
            }

            // assigning every token to different variable and creating variable result that will later be updated
            double firstNumber = double.Parse(tokensOfInput[0]);
            string wantedOperation = tokensOfInput[1];
            double secondNumber = double.Parse(tokensOfInput[2]);
            double result = 0;

            // checking users wanted operation and updating history
            switch (wantedOperation)
            {
                case "+":
                    result = firstNumber + secondNumber;
                    Console.WriteLine($"Result: {result}");
                    SavingHistory(userInput, result);
                    break;
                case "-":
                    result = firstNumber - secondNumber;
                    Console.WriteLine($"Result: {result}");
                    SavingHistory(userInput, result);
                    break;
                case "*":
                    result = firstNumber * secondNumber;
                    Console.WriteLine($"Result: {result}");
                    SavingHistory(userInput, result);
                    break;
                case "/":
                    if (secondNumber != 0)
                    {
                        result = firstNumber / secondNumber;
                        Console.WriteLine($"Result: {result}");
                        SavingHistory(userInput, result);
                    }
                    else
                    {
                        Console.WriteLine("Error: You can not divide by zero.");
                    }
                    break;
                default:
                    Console.WriteLine("Invalid operator. Use +, -, *, or /.");
                    break;
            }
        }

        catch (FormatException)
        {
            Console.WriteLine("Invalid input format. Please use the format: operand (space) operator (space) operand.");
        }
    }
    static void SecondOption()
    {
        Console.Write("Please enter a number to calculate its square root: ");
        string userInput = Console.ReadLine();

        try
        {
            double userNumber = double.Parse(userInput);

            // if number is less than zero, no square root
            if (userNumber < 0)
            {
                throw new ArgumentOutOfRangeException("No calculations of the square root of a negative number.");
            }

            //calculating and saving to history
            double result = Math.Sqrt(userNumber);
            Console.WriteLine($"Square Root: √{userNumber} = {result}");
            SavingHistory($"√{userNumber}", result);
        }
        catch (FormatException)
        {
            Console.WriteLine("You can not enter letters. Please enter a number.");
        }
        catch (ArgumentOutOfRangeException ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
        }
    }
    static void ThirdOption()
    {
        CheckingHistoryExists();
        var history = JsonConvert.DeserializeObject<List<HistoryOfCalculation>>(File.ReadAllText(jsonPath));

        if (history == null || history.Count == 0)
        {
            Console.WriteLine("No history available.");
        }
        else
        {
            Console.WriteLine("Calculation History:");
            int count = 1;
            foreach (var record in history)
            {
                Console.WriteLine($"{count}: {record.Expression} = {record.Result}");
                count++;
            }
        }
    }
    static void SavingHistory(string expression, double result)
    {
        CheckingHistoryExists();
        var history = JsonConvert.DeserializeObject<List<HistoryOfCalculation>>(File.ReadAllText(jsonPath))
                      ?? new List<HistoryOfCalculation>();
        history.Add(new HistoryOfCalculation
        {
            Expression = expression,
            Result = result
        });
        File.WriteAllText(jsonPath, JsonConvert.SerializeObject(history, Formatting.Indented));
    }
    public class HistoryOfCalculation
    {
        public string Expression { get; set; }
        public double Result { get; set; }
    }
}
